cmd/load-generator/boulder-calls.go

```diff
@@ -0,0 +1,482 @@
+	s.challSrv.addHTTPOneChallenge(chall.Token, authStr)
+
+	update := fmt.Sprintf(`{"resource":"challenge","keyAuthorization":"%s"}`, authStr)
+	requestPayload, err := s.signWithNonce("/acme/challenge", false, []byte(update), ctx.reg.signer)
+	if err != nil {
+		return err
+	}
+
+	cStarted := time.Now()
+	resp, err := s.post(chall.URI, requestPayload, ctx.ns)
+	cFinished := time.Now()
+	cState := "good"
+	defer func() { s.callLatency.Add("POST /acme/challenge/{ID}", cStarted, cFinished, cState) }()
+	if err != nil {
+		cState = "error"
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 202 {
+		cState = "error"
+		return fmt.Errorf("Unexpected error code")
+	}
+	// Sit and spin until status valid or invalid
+	ident := ""
+	for i := 0; i < 3; i++ {
+		aStarted := time.Now()
+		resp, err = s.get(fmt.Sprintf("%s/acme/authz/%s", s.apiBase, authz.ID))
+		aFinished := time.Now()
+		aState := "good"
+		defer func() { s.callLatency.Add("GET /acme/authz/{ID}", aStarted, aFinished, aState) }()
+		if err != nil {
+			aState = "error"
+			return fmt.Errorf("/acme/authz bad response: %s", err)
+		}
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			// just fail
+			aState = "error"
+			return err
+		}
+		var newAuthz core.Authorization
+		err = json.Unmarshal(body, &newAuthz)
+		if err != nil {
+			aState = "error"
+			return fmt.Errorf("/acme/authz bad response: %s", body)
+		}
+		if newAuthz.Status == "valid" {
+			ident = newAuthz.Identifier.Value
+			break
+		}
+		if newAuthz.Status == "invalid" {
+			return fmt.Errorf("HTTP-01 challenge invalid: %s", string(body))
+		}
+		time.Sleep(3 * time.Second) // XXX: Mimics certbot behaviour
+	}
+	if ident == "" {
+		return errors.New("HTTP-01 challenge validation timed out")
+	}
+
+	ctx.finalizedAuthz = append(ctx.finalizedAuthz, ident)
+	return nil
+}
+
+func solveTLSOne(s *State, ctx *context) error {
+	if len(ctx.pendingAuthz) == 0 {
+		return errors.New("no pending authorizations to complete")
+	}
+	authz := getPending(ctx)
+	var chall *core.Challenge
+	for _, c := range authz.Challenges {
+		if c.Type == "tls-sni-01" {
+			chall = &c
+			break
+		}
+	}
+	if chall == nil {
+		return errors.New("no http-01 challenges to complete")
+	}
+
+	jwk := &jose.JsonWebKey{Key: &ctx.reg.key.PublicKey}
+	thumbprint, err := jwk.Thumbprint(crypto.SHA256)
+	if err != nil {
+		return err
+	}
+	authStr := fmt.Sprintf("%s.%s", chall.Token, base64.RawURLEncoding.EncodeToString(thumbprint))
+
+	update := fmt.Sprintf(`{"resource":"challenge","keyAuthorization":"%s"}`, authStr)
+	requestPayload, err := s.signWithNonce("/acme/challenge", false, []byte(update), ctx.reg.signer)
+	if err != nil {
+		return err
+	}
+
+	cStarted := time.Now()
+	resp, err := s.post(chall.URI, requestPayload, ctx.ns)
+	cFinished := time.Now()
+	cState := "good"
+	defer func() { s.callLatency.Add("POST /acme/challenge/{ID}", cStarted, cFinished, cState) }()
```

Comment:
I've kept it as `{ID}` as it makes clumping all of the stuff together for charting with the Python script much easier. We could keep the actual ID and do the truncating in the script but I'm not sure it adds much value.